packages/compass-data-modeling/src/services/data-model-storage.ts

```diff
@@ -1,4 +1,96 @@
 import { z } from '@mongodb-js/compass-user-data';
+import type { MongoDBJSONSchema } from 'mongodb-schema';
+
+export const RelationshipSideSchema = z.object({
+  ns: z.string(),
+  cardinality: z.number(),
+  fields: z.array(z.string()),
+});
+
+export type RelationshipSide = z.output<typeof RelationshipSideSchema>;
+
+export const RelationshipSchema = z.object({
+  id: z.string(),
+  relationship: z.tuple([RelationshipSideSchema, RelationshipSideSchema]),
+  isInferred: z.boolean(),
+});
+
+export type Relationship = z.output<typeof RelationshipSchema>;
+
+export const StaticModelSchema = z.object({
+  collections: z.array(
+    z.object({
+      ns: z.string(),
+      jsonSchema: z.unknown(), // skipped for simplicity
+      indexes: z.array(z.record(z.unknown())),
+      shardKey: z.record(z.unknown()).optional(),
+      displayPosition: z.tuple([z.number(), z.number()]),
+    })
+  ),
+  relationships: z.array(RelationshipSchema),
+});
+
+export type StaticModel = Omit<
+  z.output<typeof StaticModelSchema>,
+  'collections'
+> & {
+  collections: Array<
+    Omit<
+      z.output<typeof StaticModelSchema>['collections'][number],
+      'jsonSchema'
+    > & {
+      jsonSchema: MongoDBJSONSchema;
+    }
+  >;
+};
+
+export const EditSchema = z.discriminatedUnion('type', [
+  z.object({
+    type: z.literal('SetModel'),
+    id: z.string(),
+    timestamp: z.string(),
+    model: StaticModelSchema,
+  }),
+  z.object({
+    type: z.literal('AddRelationship'),
+    id: z.string(),
```

Comment:
`id` `type` will be shared with almost every `Edit`, should we extract it out and then do `z.interection` on that and this `EditSchema`?